alacritty_terminal/src/vi_mode.rs

```diff
@@ -756,4 +746,40 @@ mod tests {
         cursor = cursor.motion(&mut term, ViMotion::WordLeft);
         assert_eq!(cursor.point, Point::new(Line(0), Column(0)));
     }
+
+    fn history_term() -> Term<()> {
+        let mut term = term();
+        term.grid_mut().clear_viewport();
+        term
+    }
+
+    #[test]
+    fn scroll_simple() {
+        let term = history_term();
+
+        let mut cursor = ViModeCursor::new(Point::new(Line(0), Column(0)));
+
+        cursor = cursor.scroll(&term, -1);
+        assert_eq!(cursor.point, Point::new(Line(1), Column(0)));
+
+        cursor = cursor.scroll(&term, 1);
+        assert_eq!(cursor.point, Point::new(Line(0), Column(0)));
+
+        cursor = cursor.scroll(&term, 1);
+        assert_eq!(cursor.point, Point::new(Line(-1), Column(0)));
+    }
+
+    #[test]
+    fn scroll_over() {
+        let term = history_term();
+        let overlines = (term.total_lines() * 2) as i32;
+
+        let mut cursor = ViModeCursor::new(Point::new(Line(0), Column(0)));
+
+        cursor = cursor.scroll(&term, -overlines);
+        assert_eq!(cursor.point, Point::new(Line(term.screen_lines() as i32 - 1), Column(0)));
+
+        cursor = cursor.scroll(&term, overlines);
+        assert_eq!(cursor.point, Point::new(Line(-(term.history_size() as i32)), Column(0)));
+    }
```

Comment:
I'm not sure I like the whole `overlines` thing here. It makes stuff somewhat complicated. Especially because the cursor position is preserved between the two tests so one not only needs to make sure that the second test overshoots from the 0-0 point, but from the end position of the previous test. Making it possible for the second test to fail by running the first part just by picking the wrong `overlines` value.

I think being more explicit and inlining `overlines` to be specific to the thing we want to test right at that moment would clear things up a bit. It might also make sense to reset the cursor position between tests. That way by writing things out a bit more explicitly it should be easier to understand what we're trying to validate with the test.

Reply:
OK. I rewrite this function and separate it into two functions to be explicit.